### ISSUE TITLE
BounDDoULoss for 3D dataset

### ISSUE THREAD

**Issue by AN-AN-369** [sunfan-bvb/BoundaryDoULoss#1]
Thank you for your contributions in the article! May I use BounDDoULoss for 3D dataset?"

**Comment by sunfan-bvb** [comment]
Of course this is a good idea! But since our alpha is calculated based on the boundary length and size, it may need to be converted to surface area and volume in the 3D dataset, or slice the 3D result to calculate the loss, thanks for using and looking forward to your results!

**Comment by Overflowu1** [comment]
I have modified the Bou as follows, and it can run successfully in 3D tasks, but I don't know if this change is reasonable or conforms to the theory. Could you please help me have a look?
class BoundaryDoULoss(nn.Module):
    def __init__(self, n_classes):
        super(BoundaryDoULoss, self).__init__()
        self.n_classes = n_classes

    def _one_hot_encoder(self, input_tensor):
        tensor_list = []
        for i in range(self.n_classes):
            temp_prob = input_tensor == i
            tensor_list.append(temp_prob.unsqueeze(1))
        output_tensor = torch.cat(tensor_list, dim=1)
        return output_tensor.float()

    def _adaptive_size(self, score, target):
        kernel = torch.Tensor([[[[0, 0, 0], [0, 1, 0], [0, 0, 0]],
                                [[0, 1, 0], [1, 1, 1], [0, 1, 0]],
                                [[0, 0, 0], [0, 1, 0], [0, 0, 0]]]])
        padding_out = torch.zeros((target.shape[0], target.shape[-3] + 2, target.shape[-2] + 2, target.shape[-1] + 2))
        padding_out[:, 1:-1, 1:-1, 1:-1] = target
        h, w, d = 3, 3, 3

        Y = torch.zeros((padding_out.shape[0], padding_out.shape[1] - d + 1, padding_out.shape[2] - h + 1,
                         padding_out.shape[3] - w + 1)).cuda()
        for i in range(Y.shape[0]):
            Y[i, :, :,:] = torch.conv3d(target[i].unsqueeze(0),kernel.unsqueeze(0).cuda(), padding=1)
        Y = Y * target
        Y[Y == 5] = 0
        C = torch.count_nonzero(Y)
        S = torch.count_nonzero(target)
        smooth = 1e-5
        alpha = 1 - (C + smooth) / (S + smooth)
        alpha = 2 * alpha - 1

        intersect = torch.sum(score * target)
        y_sum = torch.sum(target * target)
        z_sum = torch.sum(score * score)
        alpha = min(alpha, 0.8)
        loss = (z_sum + y_sum - 2 * intersect + smooth) / (z_sum + y_sum - (1 + alpha) * intersect + smooth)

        return loss

    def forward(self, inputs, target):
        inputs = torch.softmax(inputs, dim=1)
        target = self._one_hot_encoder(target)

        assert inputs.size() == target.size(), 'predict {} & target {} shape do not match'.format(inputs.size(),
                                                                                                  target.size())

        loss = 0.0
        for i in range(0, self.n_classes):
            loss += self._adaptive_size(inputs[:, i], target[:, i])
        return loss / self.n_classes

**Comment by sunfan-bvb** [comment]
Hi, in theory this code does have a bit of a problem. Calculating the side length and area of an object on a 2D image looks like this:
![calcuate](https://github.com/sunfan-bvb/BoundaryDoULoss/assets/55881602/79b6389a-e4ae-4b1b-b57f-409c69237093)
So in 3D, based on the kernel you are using, you should modify `Y[Y == 5] = 0` to `Y[Y == 7] = 0`. Since I am not familiar with 3D convolution, you can refer to the above diagram to modify the code further.

**Comment by Overflowu1** [comment]
Thanks for the advice.Since Bou doesn't perform as well as dice or cross-entropy plus dice in 3D tasks, I'm wondering if there's something wrong with my application. I'll make changes to the code first and come back if I have any questions

**Comment by Om-Pandey** [comment]
Would it be a better way to use the same way of 2D but in 3D take dimensions 2 at a time and average by 3?

**Comment by sunfan-bvb** [comment]
I'm not sure. Just give it a try!

**Comment by Halleyawoo** [comment]
@Overflowu7 Hello! Have you successfully modified the loss function for three-dimensional data and achieved good results? Could you please share how you adjusted it for three-dimensional data? Looking forward to your response. Thank you!

**Comment by innat** [comment]
@sunfan-bvb 
Based on the above discussion, I tried to implement its 3d version as follows, though I still need to check its effectiveness. In one my test run, it did give promising results but had to combine with other losses (dice_ce + bdou).

```python
import keras 
from keras import ops

def get_3d_cross_kernel():
    k = np.zeros((3, 3, 3), dtype="float32")
    k[1, 1, 1] = 1
    k[0, 1, 1] = k[2, 1, 1] = 1
    k[1, 0, 1] = k[1, 2, 1] = 1
    k[1, 1, 0] = k[1, 1, 2] = 1
    return k[..., None, None]  # (3,3,3,1,1)

class BoundaryDoULoss3D(keras.losses.Loss):
    def __init__(self, num_classes, ignore_class_id=None, smooth=1e-5, name="boundary_dou"):
        super().__init__(name=name)
        self.num_classes = num_classes
        self.ignore_class_id = ignore_class_id
        self.smooth = smooth
        self.kernel = get_3d_cross_kernel()

    def call(self, y_true, y_pred):
        y_true = ops.cast(y_true, "int32")
        
        # One-hot
        y_true_oh = ops.one_hot(y_true[..., 0], self.num_classes)

        class_indices = [
            i for i in range(self.num_classes) 
            if i != self.ignore_class_id
        ]
        y_true_oh = ops.take(y_true_oh, class_indices, axis=-1)
        y_pred = ops.take(y_pred, class_indices, axis=-1)

        if self.ignore_class_id is not None:
            valid_mask = ops.cast(
                ops.not_equal(y_true[..., 0], self.ignore_class_id),
                "float32"
            )[..., None]
            y_true_oh *= valid_mask
            y_pred *= valid_mask

        # Boundary detection
        shape = ops.shape(y_true_oh)  # (B,D,H,W,C)

        y_true_flat = ops.reshape(
            ops.transpose(y_true_oh, (0, 4, 1, 2, 3)),
            (-1, shape[1], shape[2], shape[3], 1)
        )

        conv_res = ops.conv(y_true_flat, self.kernel, padding="same")
        conv_res = ops.reshape(
            conv_res, (shape[0], shape[4], shape[1], shape[2], shape[3])
        )
        conv_res = ops.transpose(conv_res, (0, 2, 3, 4, 1))

        boundary_mask = conv_res * y_true_oh
        boundary_mask = ops.where(
            ops.equal(boundary_mask, 7.0),
            0.0,
            boundary_mask
        )

        # Adaptive alpha (per class)
        C = ops.sum(ops.cast(boundary_mask > 0, "float32"), axis=(1,2,3))
        S = ops.sum(y_true_oh, axis=(1,2,3))

        alpha = 1.0 - (C + self.smooth) / (S + self.smooth)
        alpha = 2.0 * alpha - 1.0
        alpha = ops.minimum(alpha, 0.8)

        # DoU loss
        intersect = ops.sum(y_pred * y_true_oh, axis=(1,2,3))
        z_sum = ops.sum(y_pred * y_pred, axis=(1,2,3))
        y_sum = ops.sum(y_true_oh * y_true_oh, axis=(1,2,3))

        numerator = z_sum + y_sum - 2.0 * intersect + self.smooth
        denominator = z_sum + y_sum - (1.0 + alpha) * intersect + self.smooth
        loss = numerator / denominator  # (B, C)
        return ops.mean(loss)
```

(This loss is written specificly for sparse label where [0, 1, 2] are true labels and 2 is invalid region.)
